Add API Reference (#2)
tyntec/botbuilder-adapter-tyntec-whatsapp: docs/Activity.md

# `interface Activity`

`Activity` is a Microsoft Bot Framework representation of actions made by
conversation participants.

> The full specification of the `Activity` interface can be found in the

Reviewer: ```suggestion
You can find the full specification of the `Activity` interface in the
```

Author: Changed now: https://github.com/yet-another-radek/botbuilder-adapter-tyntec-whatsapp/compare/daa8a90..b6a7720#diff-e17fbb320b84961a419f10a0a4a14668d276f50ebcaa57ac4cf1f5d4197cd4e6R6

Microsoft Bot Framework SDK documentation at https://docs.microsoft.com/en-us/azure/bot-service/index-bf-sdk

> and in the Microsoft Bot Framework SDK specs at https://github.com/microsoft/botframework-sdk/tree/main/specs.

Reviewer: ```suggestion
and in Microsoft Bot Framework SDK specs at https://github.com/microsoft/botframework-sdk/tree/main/specs.
```

Author: ...and the _the_ before _Microsoft Bot Framework SDK documentation_ in the same paragraph is ok?

Reviewer: TL;DR - Languages are mess and they rules don't exist.

In some degree, yes. Documentation is both countable and uncountable (thanks, English) and it this case we speak about a specific thing rather than abstract case. That is why I kept it there (it also sounded better to my ear while reading it with thick Scottish accent).

Specs on the other hand sound more like plural (specifications) than singular specification, so went with the no article in front of plural. However, the article can technically still stay if the Microsoft Bot Framework SDK specs is a particular name or something really specific.

Author: Changed now: https://github.com/yet-another-radek/botbuilder-adapter-tyntec-whatsapp/compare/e36cc18..2025655

However, at the moment, Tyntec WhatsApp Adapter supports only a limited subset

> of activities. This means, that there are stricter requirements for activities

Reviewer: ```suggestion
of activities. This means, that both activities passed to the adapter and adapter created activities have stricter requirements.  
```

Author: Changed now: https://github.com/yet-another-radek/botbuilder-adapter-tyntec-whatsapp/compare/daa8a90..b6a7720#diff-e17fbb320b84961a419f10a0a4a14668d276f50ebcaa57ac4cf1f5d4197cd4e6R11-R12

> passed to the adapter as well as for activities created by the adapter.

Reviewer: ```suggestion
```

Author: Changed now: https://github.com/yet-another-radek/botbuilder-adapter-tyntec-whatsapp/compare/daa8a90..b6a7720#diff-e17fbb320b84961a419f10a0a4a14668d276f50ebcaa57ac4cf1f5d4197cd4e6R11-R12

The only supported activity type is a [WhatsApp message](#whatsapp-message-activity)
(`type: "message", channelId: "whatsapp"`).

## WhatsApp Message Activity

> The sender of the activity (`from`) MUST be defined. The channel account ID

Author: @ridlees, I've just camel-cased MUST, MUST NOT and MAY in the Activity.md to mimic the official Microsoft specs and RFCs... https://github.com/yet-another-radek/botbuilder-adapter-tyntec-whatsapp/compare/7d842aa..96d847d

> The sender of the activity (`from`) must be defined. The channel account ID

Reviewer: Do you think we could use bullet lists for these paragraphs? 

Must be defined:
- `from` is sender of the activity
- `from.id` is the WhatsApp phone number of the sender

Can be defined:
- `from.name` can be the display name of the sender

Ignored/Should not be defined: 
- `entities` -  the metadata pertaining objects
- `replyToId` - the ID of prior activity
- `textFormat` -  the text format

Or in format of the top level objects with nested bullet lists

-  `from` must be defined
    - `from.id` is  the sender's WhatsApp phone number
    - `from.name` can be the name of the sender
-  `recipient` is defined in channel activities
    - ` recipient.id` is the recipient's WhatsApp phone number

Author: I tried to stick to the format of the Microsoft specs here: https://github.com/microsoft/botframework-sdk/blob/main/specs/botframework-activity/botframework-activity.md Do you still think it's better to use the lists? I don't have a preference here.

Reviewer: I think list generally make us write less words and thus prevent big chunks of text (which are, sadly, discouraging for the common reader). So, in general, i think they are better in terms of readability.

However, the emphasis with (MUST MUST NOT MAY) works on similar level (I expect the reader to scan these parts rather than read thoroughly (yes, it is reference, but you mostly care about the key pair (from and MUST, deliveryMode and MUST NOT)). And if this mimics the way MS is doing documentation, it can seem more familiar to the active user, which can be better call here.

So, I would personally go with bullet lists here because:
- They improve readability (especially if we are giving bunch of "key-pairs") 
- They are easier to maintain
- They give less space to confusion (but also have less space to explain something)

Author: What about this? https://github.com/yet-another-radek/botbuilder-adapter-tyntec-whatsapp/compare/2025655..f66608b

activity exists (`conversation.id`) is set to the WhatsApp phone number of the
sender, the account name (`conversation.name`) may be set to the display name
of the sender and the account Is Group (`conversation.isGroup`) is set to
`false`. In case of bot activities, the account ID (`conversation.id`) must be
set to the WhatsApp phone number of the recipient.

The prior activity to which the activity is a reply (`replyToId`), the metadata
pertaining objects (`entities`), the text format (`textFormat`), the language
code of the text content (`locale`), the instructions how to present
attachments (`attachmentLayout`), the interactive actions that may be displayed
to the user (`suggestedActions`), the time at which the activity should be
considered to be expired (`expiration`), the relative importance
(`importance`), the terms that language processing systems can listen for
(`listenFor`) and the programmatic action accompanying the user request
(`semanticAction`) are ignored.

The text-to-speech content of the activity (`speak`), the indication of whether
the sender is anticipating a response (`inputHint`) and the alternate delivery
paths (`deliveryMode`) must not be defined.

> Only supported WhatsApp messages are [audio](#whatsapp-audio-message-activity),

Reviewer: ```suggestion
The supported WhatsApp messages are [audio](#whatsapp-audio-message-activity),
```

Author: Changed now: https://github.com/yet-another-radek/botbuilder-adapter-tyntec-whatsapp/compare/daa8a90..b6a7720#diff-e17fbb320b84961a419f10a0a4a14668d276f50ebcaa57ac4cf1f5d4197cd4e6R50

[document](#whatsapp-document-message-activity),
[image](#whatsapp-image-message-activity),
[sticker](#whatsapp-sticker-message-activity),
[template](#whatsapp-template-message-activity),
[text](#whatsapp-text-message-activity) and
[video](#whatsapp-video-message-activity) messages.

### WhatsApp Audio Message Activity

The channel data of the activity (`channelData`) must be defined. A content
type (`channelData.contentType`) must be set to `"audio"`. A message template
(`channelData.template`) must not be defined.

The text content of the activity (`text`) must not be defined.

Exactly one attachment to the activity (`attachments`) must be defined. The
content URL (`attachments[i].contentUrl`) must be defined. The content
(`attachments[i].content`) and the thumbnail URL
(`attachments[i].thumbnailUrl`) must not be defined.

> An example of a WhatsApp audio message activity looks like this:

Reviewer: ```suggestion
The WhatsApp audio message activity example:
```
Just a suggestion to shorten it. Maybe even Audio message example could work here / or code example.

Author: I don't see a difference so why not. But... Why _the_? Should it be _A WhatsApp audio message activity example_?

Reviewer: You are right, It should be :D . The article got there since I was playing with multiple versions. Maybe we can omit the word activity and keep just WhatsApp audio message example.

Author: Changed now: https://github.com/yet-another-radek/botbuilder-adapter-tyntec-whatsapp/compare/f66608b..9a5f449

```javascript
activity === {
    type: "message",
    channelId: "whatsapp",
    id: "77185196-664a-43ec-b14a-fe97036c697e",
    timestamp: new Date("2019-06-26T09:41:00.000Z"),
    from: {
        id: "+1233423454"
    },
    recipient: {
        id: "545345345"
    },
    conversation: {
        id: "+1233423454",
        isGroup: false,
        name: "John Doe"
    },
    channelData: {
        contentType: "audio"
    },
    serviceUrl: "https://api.tyntec.com/conversations/v3/messages",
    attachments: [
        {

>             contentType: "audio/aac",

Author: @ridlees, so, the tyntec API understands all MIME that are listed here: https://developers.facebook.com/docs/whatsapp/api/media/ ? Nothing more, nothing less?

Author: @ridlees, added now: https://github.com/yet-another-radek/botbuilder-adapter-tyntec-whatsapp/compare/a1d10e5..66a4a37

    text: "A simple text message"
}
```

### WhatsApp Video Message Activity

The channel data of the activity (`channelData`) must be defined. A content
type (`channelData.contentType`) must be set to `"video"`. A message template
(`channelData.template`) must not be defined.

The text content of the activity (`text`) must not be defined.

Exactly one attachment to the activity (`attachments`) must be defined. The
content URL (`attachments[i].contentUrl`) must be defined. The content
(`attachments[i].content`) and the thumbnail URL
(`attachments[i].thumbnailUrl`) must not be defined.

An example of a WhatsApp video message activity looks like this:

```javascript
activity === {
    type: "message",
    channelId: "whatsapp",
    id: "77185196-664a-43ec-b14a-fe97036c697e",
    timestamp: new Date("2019-06-26T09:41:00.000Z"),
    from: {
        id: "+1233423454"
    },
    recipient: {
        id: "545345345"
    },
    conversation: {
        id: "+1233423454",
        isGroup: false,
        name: "John Doe"
    },
    channelData: {
        contentType: "video"
    },
    serviceUrl: "https://api.tyntec.com/conversations/v3/messages",

>     text: "A video caption",

Reviewer: ```suggestion
```

On line 347 we say: _The text content of the activity (`text`) must not be defined._ 

Technically, the API ignores unexpected fields. So this is only for consistency.

Author: Damn, I've at least triple-checked that :facepalm: Fixed now: https://github.com/yet-another-radek/botbuilder-adapter-tyntec-whatsapp/compare/492fbba9..daa8a90
